src/GitHub.VisualStudio/Commands/SyncSubmodulesCommand.cs

```diff
@@ -0,0 +1,100 @@
+    {
+        static readonly ILogger log = LogManager.ForContext<SyncSubmodulesCommand>();
+
+        readonly Lazy<IPullRequestService> lazyPullRequestService;
+        readonly Lazy<IStatusBarNotificationService> lazyStatusBarNotificationService;
+        readonly Lazy<IVSGitExt> lazyVSGitExt;
+
+        [ImportingConstructor]
+        protected SyncSubmodulesCommand(
+            Lazy<IPullRequestService> pullRequestService,
+            Lazy<IStatusBarNotificationService> statusBarNotificationService,
+            Lazy<IVSGitExt> gitExt)
+            : base(CommandSet, CommandId)
+        {
+            lazyPullRequestService = pullRequestService;
+            lazyStatusBarNotificationService = statusBarNotificationService;
+            lazyVSGitExt = gitExt;
+        }
+
+        /// <summary>
+        /// Gets the GUID of the group the command belongs to.
+        /// </summary>
+        public static readonly Guid CommandSet = Guids.guidGitHubCmdSet;
+
+        /// <summary>
+        /// Gets the numeric identifier of the command.
+        /// </summary>
+        public const int CommandId = PkgCmdIDList.syncSubmodulesCommand;
+
+        /// <summary>
+        /// Syncs submodules.
+        /// </summary>
+        public override async Task Execute()
+        {
+            try
+            {
+                var complete = await SyncSubmodules();
+            }
+            catch (Exception ex)
+            {
+                log.Error(ex, "Error syncing submodules");
+                lazyStatusBarNotificationService.Value.ShowMessage("Error syncing submodules");
+            }
+        }
+
+        /// <summary>
+        /// Sync submodules in local repository.
+        /// </summary>
+        /// <returns>Tuple with bool that is true if command completed successfully and string with
+        /// output from sync submodules Git command.</returns>
+        public async Task<Tuple<bool, string>> SyncSubmodules()
+        {
+            var pullRequestService = lazyPullRequestService.Value;
+            var statusBarNotificationService = lazyStatusBarNotificationService.Value;
+            var gitExt = lazyVSGitExt.Value;
+
+            var repository = gitExt.ActiveRepositories.FirstOrDefault();
+            if (repository == null)
+            {
+                statusBarNotificationService.ShowMessage("No local Git repository");
+                return new Tuple<bool, string>(true, "No local Git repository");
+            }
+
+            var writer = new StringWriter(CultureInfo.CurrentCulture);
+            var complete = await pullRequestService.SyncSubmodules(repository, line =>
+            {
+                writer.WriteLine(line);
+                statusBarNotificationService.ShowMessage(line);
+            });
+
+            if (!complete)
+            {
+                statusBarNotificationService.ShowMessage("Failed to sync submodules." + Environment.NewLine + writer);
```

Comment:
Is it normal for VS commands to write directly to the status bar? Is there no concept of "output" for commands, which e.g. shows in the command pane?

Reply:
I think the status bar is what commands tend to use as their ambient UI. There isn't really any concept of a native UI or output for a command.

One possibility would be to pass in a progress object that the command could report back to. This could also sidestep the issue of returning results.

I wanted to start off with a simple refactoring and and avoid any extra implementation detail. We can revisit next time this command is touched (if we decide to surface it somewhere else).